apps/emqx_retainer/src/emqx_retainer_dispatcher.erl

```diff
@@ -174,9 +192,74 @@ cast(Msg) ->
 worker() ->
     gproc_pool:pick_worker(?POOL, self()).
 
-run({M, F, A}) ->
-    erlang:apply(M, F, A);
-run({F, A}) when is_function(F), is_list(A) ->
-    erlang:apply(F, A);
-run(Fun) when is_function(Fun) ->
-    Fun().
+-spec dispatch(context(), pid(), topic(), cursor(), limiter()) -> {ok, limiter()}.
+dispatch(Context, Pid, Topic, Cursor, Limiter) ->
+    Mod = emqx_retainer:get_backend_module(),
+    case Cursor =/= undefined orelse emqx_topic:wildcard(Topic) of
+        false ->
+            {ok, Result} = Mod:read_message(Context, Topic),
+            deliver(Result, Context, Pid, Topic, undefined, Limiter);
+        true  ->
+            {ok, Result, NewCursor} =  Mod:match_messages(Context, Topic, Cursor),
+            deliver(Result, Context, Pid, Topic, NewCursor, Limiter)
+    end.
+
+-spec deliver(list(emqx_types:message()), context(), pid(), topic(), cursor(), limiter()) -> {ok, limiter()}.
+deliver([], _Context, _Pid, _Topic, undefined, Limiter) ->
+    {ok, Limiter};
+
+deliver([], Context, Pid, Topic, Cursor, Limiter) ->
+    dispatch(Context, Pid, Topic, Cursor, Limiter);
+
+deliver(Result, Context, Pid, Topic, Cursor, Limiter) ->
+    case erlang:is_process_alive(Pid) of
+        false ->
+            {ok, Limiter};
+        _ ->
+            DeliverNum = emqx:get_config([retainer, flow_control, batch_deliver_number]),
+            case DeliverNum of
+                0 ->
+                    do_deliver(Result, Pid, Topic),
+                    {ok, Limiter};
+                _ ->
+                    case do_deliver(Result, DeliverNum, Pid, Topic, Limiter) of
+                        {ok, Limiter2} ->
+                            deliver([], Context, Pid, Topic, Cursor, Limiter2);
+                        {drop, Limiter2} ->
+                            {ok, Limiter2}
+                    end
+            end
+    end.
+
+do_deliver([], _DeliverNum, _Pid, _Topic, Limiter) ->
+    {ok, Limiter};
+
+do_deliver(Msgs, DeliverNum, Pid, Topic, Limiter) ->
+    {Num, ToDelivers, Msgs2} = safe_split(DeliverNum, Msgs),
+    case emqx_htb_limiter:consume(Num, Limiter) of
+        {ok, Limiter2} ->
+            do_deliver(ToDelivers, Pid, Topic),
+            do_deliver(Msgs2, DeliverNum, Pid, Topic, Limiter2);
+        {drop, _} = Drop ->
+            ?SLOG(error, #{msg => "the retainer deliver failed because the required quota could not be obtained"}),
```

Comment:
It looks like we'll just drop the delivery of Retaine messages that are limited due to flow control?

Reply:
yes, drop is a kind of emqx_limiter failure strategy, so it needs to be handled  here